### PR TITLE
feat: validator for non null and non empty parameter

### LINKED ISSUE #55: Need a new parameter validation function
We need to validate parameters based on their value

Currently we use the following to validate null values:

```php
HeaderParam::init('key', '')->required();
```

But we need another function that does the same but also check for empty values like:

```php
HeaderParam::init('key', '')->requiredNonEmpty();
```

### PR DESCRIPTION
**Pull request by asadali214**
## What
This PR adds implementation for new validator function from ParamInterface to check if a parameter's value is non null or non empty while authentication initialization.

## Why
We need to fail the validation if an authentication parameter value is set to an empty string.

Closes #55

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause a breaking change)
- [X] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
N/A

## Breaking change
N/A

## Testing
Unit tests are added to test the implementation of new functions

## Checklist
- [X] My code follows the coding conventions
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added new unit tests

### CODE REVIEW

File: src/Request/Parameters/Parameter.php [hunk 1]
```diff
@@ -73,6 +74,18 @@ public function required(): self
         return $this;
     }
 
+    /**
+     * Marks the value of the parameter as required + non-empty and
+     * throws an exception on validate if the value is missing.
+     */
+    public function requiredNonEmpty(): self
+    {
+        if (empty($this->value)) {
```
Comment (sufyankhanrao) [inline src/Request/Parameters/Parameter.php:83]: Does this handle null value?

File: tests/AuthenticationTest.php [hunk 2]
```diff
@@ -85,6 +95,19 @@ public function testHeaderWithMissingFieldOrQueryAuth()
         );
     }
 
+    public function testHeaderWithEmptyFieldOrQueryAuth()
```
Comment (sufyankhanrao) [inline tests/AuthenticationTest.php:98]: I'd recommend adding another test case for `and` auth for having one of the auth param's value as empty and then asserting the exception. This will enforce the required behavior for `and` auth.

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]
Above mentioned suggestions need your attention.

File: src/Request/Parameters/Parameter.php (hunk 1 above)
Comment (asadali214) [reply src/Request/Parameters/Parameter.php:83]: yes it does!

File: tests/AuthenticationTest.php (hunk 2 above)
Comment (asadali214) [reply tests/AuthenticationTest.php:98]: done

File: src/Request/Parameters/Parameter.php (hunk 1 above)
Comment (sufyankhanrao) [reply src/Request/Parameters/Parameter.php:83]: Have we added the test case for this i.e. null param value?

Comment (asadali214) [reply src/Request/Parameters/Parameter.php:83]: yes its covered with both `null` and `""` values

File: src/Request/Parameters/FormParam.php [hunk 3]
```diff
@@ -30,9 +30,6 @@ private function __construct(string $key, $value)
 
     /**
      * Sets encoding header with the key and value provided.
-     *
-     * @param string $key
-     * @param string $value
      */
```
Comment (sufyankhanrao) [inline src/Request/Parameters/FormParam.php:33]: Question: Why the parameter documentation is removed in this PR?

File: src/Utils/CoreHelper.php [hunk 4]
```diff
@@ -84,6 +84,17 @@ public static function isAssociative(array $array): bool
         return false;
     }
 
+    /**
+     * Check if provided value is null or empty
+     */
+    public static function isNullOrEmpty($value): bool
```
Comment (sufyankhanrao) [inline src/Utils/CoreHelper.php:90]: Nits, Non-blocker: Method docblock should include the parameter documentation.

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]

File: src/Request/Parameters/FormParam.php (hunk 3 above)
Comment (asadali214) [reply src/Request/Parameters/FormParam.php:33]: This was not required in the first place

File: composer.json [hunk 5]
```diff
@@ -16,7 +16,7 @@
     "ext-curl": "*",
     "ext-dom": "*",
     "ext-libxml": "*",
-    "apimatic/core-interfaces": "~0.1.0",
+    "apimatic/core-interfaces": "dev-required-non-empty-validator",
```
Comment (asadali214) [inline composer.json:19]: This should be changed before merging the PR

**Review by sufyankhanrao — APPROVED** [review]
